### PR TITLE
initializer should support TensorShape

### LINKED ISSUE #21838: Keras Layer add_variable not taking Dimension objects
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Heavily based on example custom layer on tensorflow's keras page.
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Google Colab (Linux Ubuntu 17.10)
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**:
- **TensorFlow installed from (source or binary)**: Unsure 
- **TensorFlow version (use command below)**: 1.10.0
- **Python version**: 3.6.3
- **Bazel version (if compiling from source)**: N/A
- **GCC/Compiler version (if compiling from source)**: N/A
- **CUDA/cuDNN version**: Unsure. nvcc --version did not work.
- **GPU model and memory**: K80 GPU (Memory unsure)
- **Exact command to reproduce**:

You can collect some of this information using our environment capture script:

https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh

You can obtain the TensorFlow version with

python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"

### Describe the problem
My [question](https://stackoverflow.com/questions/51939618/what-could-cause-a-type-error-during-the-build-process-of-keras-layer) was answered on StackOverflow and a user recommended me to create a bug report here. I was creating a keras layer based on the NALU design, but I found that while on my local version of tensorflow (1.8.0) the code would run as expected, the same code would not run on Google Colab's version of tensorflow(1.10.0). The issue ended up being that tensorflow's Dimension type was not accepted in the Layer.add_variable method. It requires the dimensions to be converted to ints, which does not happen implicitly. I think that it would make sense for the method to accept Dimension objects as they are used to specify the shape of Tensors. As well, it provides backwards compatibility to previous versions of tensorflow, which do support pass Dimension objects to Layers.add_variable.

### Source code / logs
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached. Try to provide a reproducible test case that is the bare minimum necessary to generate the problem.

Source:
```
import tensorflow as tf

class NALU(tf.keras.layers.Layer):
    def __init__(self, num_outputs, **kwargs):
        self.num_outputs = num_outputs
        super(NALU, self).__init__(**kwargs)
        
    def build(self, input_shape):
        shape = tf.TensorShape((input_shape[1], self.num_outputs)).as_list()
        get = tf.keras.initializers.get
        self.W_ = self.add_variable("W_", shape=shape, initializer=get('glorot_uniform'))
        self.M_ = self.add_variable("M_", shape=shape, initializer=get('glorot_uniform'))
        self.GAM = self.add_variable("GAM", shape=shape, initializer=get('glorot_uniform')) # Gate add & multiply
        self.GM = self.add_variable("GM", shape=shape, initializer=get('glorot_uniform')) # Gate multiply
        
        super(NALU, self).build(input_shape)
    
    def call(self, x):
        gam = tf.sigmoid(tf.matmul(x, self.GAM)) # The gate
        gm = tf.sigmoid(tf.matmul(x, self.GM)) # The gate
        
        W = tf.tanh(self.W_) * tf.sigmoid(self.M_)
        add = tf.matmul(x, W)
        
        # represents positive multiplication
        m = tf.exp(
            tf.matmul(tf.log(tf.abs(x) + 1e-14), W)
        )
        
        mul = (-1 * m) * (1.0 - gm) + gm * m 
        
        y = gam * add + (1.0 - gam) * mul
        return y
    
    def compute_output_shape(self, input_shape):
        # --------------IMPORTANT LINE---------------------
        shape = tf.TensorShape(input_shape)
        #  shape = tf.TensorShape(input_shape).as_list() is required to make the function work
        # -------------------------------------------------
        shape[-1] = self.num_outputs
        return tf.TensorShape(shape)
    
    def get_config(self):
        base_config = super(NALU, self).get_config()
        base_config['num_outputs'] = self.num_outputs
        
    @classmethod
    def from_config(cls, config):
        return cls(**config)

def nalu_model():
    inp = tf.keras.layers.Input(shape=(2,))
    out = NALU(1)(inp)
    
    model = tf.keras.models.Model(inputs=inp, outputs=out)
    return model
```
Test Case:
```
import numpy as np

def create_train():
#     x_train = np.random.randint(-100, 100, size=(10000000, 2), dtype=np.int32)
    x_train = np.random.uniform(-100, 100, size=(10000000, 2))

    first_neg = np.copy(x_train)
    first_neg[:, 0] *= -1

    second_neg = np.copy(x_train)
    second_neg[:, 1] *= -1

    all_neg = x_train * -1

    x_train = np.append(x_train, first_neg)
    x_train = np.append(x_train, second_neg)
    x_train = np.append(x_train, all_neg)
    x_train = x_train.reshape((40000000, 2))

    com = x_train[:,::-1]
    x_train = np.append(x_train, com).reshape((80000000, 2))
    y_train = x_train[:, 0] * x_train[:, 1]
    
    return x_train, y_train

x_train, y_train = create_train()
x_test = np.random.randint(-1000, 1000, size=(10000, 2))
y_test = x_test[:, 0] * x_test[:, 1]

model = nalu_model()

model.compile(optimizer='RMSProp',
              loss='MSE',
              metrics=['accuracy', 'MAE'])

cb = [tf.keras.callbacks.TensorBoard(log_dir='./add_logs', histogram_freq=1, batch_size=32, write_graph=True, write_grads=True),
      tf.keras.callbacks.EarlyStopping(monitor='val_loss', min_delta=0.00000001, patience=5, verbose=2, mode='auto')]

model.fit(x_train, y_train, epochs=500, batch_size=256, shuffle=True, callbacks=cb, validation_split=0.4)
model.evaluate(x_test, y_test)
```

Error:
```
TypeError                                 Traceback (most recent call last)
<ipython-input-7-55fb94a8f3b1> in <module>()
     82 y_test = x_test[:, 0] * x_test[:, 1]
     83 
---> 84 model = nalu_model()
     85 
     86 model.compile(optimizer='RMSProp',

<ipython-input-7-55fb94a8f3b1> in nalu_model()
     48 def nalu_model():
     49     inp = tf.keras.layers.Input(shape=(2,))
---> 50     out = NALU(1)(inp)
     51 
     52     model = tf.keras.models.Model(inputs=inp, outputs=out)

/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/engine/base_layer.py in __call__(self, inputs, *args, **kwargs)
    726         if all(hasattr(x, 'shape') for x in input_list):
    727           input_shapes = nest.map_structure(lambda x: x.shape, inputs)
--> 728         self.build(input_shapes)
    729         self.built = True
    730 

<ipython-input-7-55fb94a8f3b1> in build(self, input_shape)
      9         shape = tf.TensorShape((input_shape[1], self.num_outputs))
     10         get = tf.keras.initializers.get
---> 11         self.W_ = self.add_variable("W_", shape=shape, initializer=get('glorot_uniform'))
     12         self.M_ = self.add_variable("M_", shape=shape, initializer=get('glorot_uniform'))
     13         self.GAM = self.add_variable("GAM", shape=shape, initializer=get('glorot_uniform')) # Gate add & multiply

/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/engine/base_layer.py in add_variable(self, *args, **kwargs)
    459   def add_variable(self, *args, **kwargs):
    460     """Alias for `add_weight`."""
--> 461     return self.add_weight(*args, **kwargs)
    462 
    463   def add_weight(self,

/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/engine/base_layer.py in add_weight(self, name, shape, dtype, initializer, regularizer, trainable, constraint, partitioner, use_resource, synchronization, aggregation, getter)
    563         use_resource=use_resource,
    564         synchronization=synchronization,
--> 565         aggregation=aggregation)
    566 
    567     if regularizer is not None:

/usr/local/lib/python3.6/dist-packages/tensorflow/python/training/checkpointable/base.py in _add_variable_with_custom_getter(self, name, shape, dtype, initializer, getter, overwrite, **kwargs_for_getter)
    533     new_variable = getter(
    534         name=name, shape=shape, dtype=dtype, initializer=initializer,
--> 535         **kwargs_for_getter)
    536 
    537     # If we set an initializer and the variable processed it, tracking will not

/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/engine/base_layer.py in make_variable(name, shape, dtype, initializer, partition_info, trainable, caching_device, validate_shape, constraint, use_resource, synchronization, aggregation, partitioner)
   1916       use_resource=use_resource,
   1917       synchronization=synchronization,
-> 1918       aggregation=aggregation)
   1919   return v

/usr/local/lib/python3.6/dist-packages/tensorflow/python/ops/variable_scope.py in variable(initial_value, trainable, collections, validate_shape, caching_device, name, dtype, constraint, use_resource, synchronization, aggregation)
   2441       use_resource=use_resource,
   2442       synchronization=synchronization,
-> 2443       aggregation=aggregation)
   2444 
   2445 

/usr/local/lib/python3.6/dist-packages/tensorflow/python/ops/variable_scope.py in <lambda>(**kwargs)
   2423              synchronization=VariableSynchronization.AUTO,
   2424              aggregation=VariableAggregation.NONE):
-> 2425   previous_getter = lambda **kwargs: default_variable_creator(None, **kwargs)
   2426   for getter in ops.get_default_graph()._variable_creator_stack:  # pylint: disable=protected-access
   2427     previous_getter = _make_getter(getter, previous_getter)

/usr/local/lib/python3.6/dist-packages/tensorflow/python/ops/variable_scope.py in default_variable_creator(next_creator, **kwargs)
   2393         collections=collections, validate_shape=validate_shape,
   2394         caching_device=caching_device, name=name, dtype=dtype,
-> 2395         constraint=constraint)
   2396   elif not use_resource and context.executing_eagerly():
   2397     raise RuntimeError(

/usr/local/lib/python3.6/dist-packages/tensorflow/python/ops/resource_variable_ops.py in __init__(self, initial_value, trainable, collections, validate_shape, caching_device, name, dtype, variable_def, import_scope, constraint)
    310           name=name,
    311           dtype=dtype,
--> 312           constraint=constraint)
    313 
    314   # pylint: disable=unused-argument

/usr/local/lib/python3.6/dist-packages/tensorflow/python/ops/resource_variable_ops.py in _init_from_args(self, initial_value, trainable, collections, validate_shape, caching_device, name, dtype, constraint)
    415               with ops.name_scope("Initializer"), ops.device(None):
    416                 initial_value = ops.convert_to_tensor(
--> 417                     initial_value(), name="initial_value", dtype=dtype)
    418               self._handle = _eager_safe_variable_handle(
    419                   shape=initial_value.get_shape(),

/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/engine/base_layer.py in <lambda>()
   1901         initializer = initializer(dtype=dtype)
   1902       init_val = lambda: initializer(  # pylint: disable=g-long-lambda
-> 1903           shape, dtype=dtype, partition_info=partition_info)
   1904       variable_dtype = dtype.base_dtype
   1905   if use_resource is None:

/usr/local/lib/python3.6/dist-packages/tensorflow/python/ops/init_ops.py in __call__(self, shape, dtype, partition_info)
    474       scale /= max(1., fan_out)
    475     else:
--> 476       scale /= max(1., (fan_in + fan_out) / 2.)
    477     if self.distribution == "normal" or self.distribution == "truncated_normal":
    478       # constant taken from scipy.stats.truncnorm.std(a=-2, b=2, loc=0., scale=1.)

TypeError: unsupported operand type(s) for /: 'Dimension' and 'float'
```

### PR DESCRIPTION
**Pull request by facaiy**
Fix #21838.

### CODE REVIEW

**Comment by facaiy** [comment]
@fchollet Could you take a look? Thanks

**Comment by martinwicke** [comment]
@fchollet this may be redundant if we remove TensorShape (or make it more tuple-like) and Dimension, but for now, this seems nice, no?

**Comment by facaiy** [comment]
Will TensorShape and Dimension be deprecated in the future?

**Comment by martinwicke** [comment]
We hope to make TensorShape much more like a tuple of ints, such that manual conversions like this would not be necessary any more.

**Comment by facaiy** [comment]
Sounds great if TensorShape could be more like a tuple of ints, but I'm afraid it might not be easy to make Dimension totally consistent with Int. Would there be a RFC later?

File: tensorflow/python/keras/engine/topology_test.py [hunk 1]
```diff
@@ -42,6 +42,34 @@
 
 class TopologyConstructionTest(test.TestCase):
 
+  def test_add_variable_supports_TensorShape(self):
```
Comment (fchollet) [inline tensorflow/python/keras/engine/topology_test.py]: This test appears unrelated to the CL; the CL adds support for TensorShape in initializers, and support in add_variable is just a side-effect. Please just add a unit test in the initializers tests.

**Review by fchollet — COMMENTED** [review]
That's a reasonable change, thanks for the PR.

File: tensorflow/python/keras/engine/topology_test.py (hunk 1 above)
Comment (facaiy) [reply tensorflow/python/keras/engine/topology_test.py]: Could you explain what's CL? changed line? And could you clarity what the new unit test is for? I think init_op_test has added tests for TensorShape.

**Comment by facaiy** [comment]
@fchollet Hi, could you make a clarification? Thanks.

**Comment by rmlarsen** [comment]
@facaiy "CL" is Google jargon for Change List which is our internal version of PR.

**Comment by facaiy** [comment]
Thanks for your explanation. Could you help review?

**Comment by rmlarsen** [comment]
@fchollet can you take another look, please?

**Comment by facaiy** [comment]
@fchollet can you take another look, please?

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @fchollet: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 17 days with no activity and the `awaiting review` label has been applied.

**Comment by facaiy** [comment]
@fchollet Friendly ping :-)

**Comment by facaiy** [comment]
@martinwicke Hi, Martin. Would you mind giving me a hand? I think the PR has been delayed for a long time :-) Any help will be appreciated, thanks.

**Comment by fchollet** [comment]
@facaiy sorry for the delay. As previously noted, this PR makes a change in `init_ops` but only adds a unit test about Keras layers' `add_weight`. Please add a unit test for precisely testing the change you're making, not a high-level integration test.

File: tensorflow/python/ops/init_ops_test.py [hunk 2]
```diff
@@ -58,114 +59,127 @@ def _runner(self,
...
-    tensor_shape = (9, 6, 7)
+    shape = [9, 6, 7]
     with self.cached_session():
-      self._runner(
-          init_ops.RandomUniform(minval=-1, maxval=1, seed=124),
-          tensor_shape,
-          target_mean=0.,
-          target_max=1,
-          target_min=-1)
+      for tensor_shape in [shape, tensor_shape_lib.TensorShape(shape)]:
```
Comment (facaiy) [inline tensorflow/python/ops/init_ops_test.py]: @fchollet Thanks for your feedback. I'm just a little confused, because I think new test cases of TensorShape for init_ops have been added. I have no idea of what new test case is needed, would you mind giving me more clues?

Comment (facaiy) [reply tensorflow/python/ops/init_ops_test.py]: By the way, do we need to delete the unit test abut keras layers ?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

File: tensorflow/python/ops/init_ops_test.py (hunk 2 above)
Comment (facaiy) [reply tensorflow/python/ops/init_ops_test.py]: I have deleted the unit test in keras, and only preserved those unit tests for init_ops.

**Review by martinwicke — APPROVED** [review]

**Comment by facaiy** [comment]
Hi, is there any update?

**Comment by martinwicke** [comment]
I think there may be conflicts. Sorry we let this linger. Can you resolve the conflicts?

**Comment by facaiy** [comment]
Thanks very much for your help, Martin.
